File: src/negotiation/domain/envelope.ts
/**
 * KNP/1.0 Negotiation Envelope（子规范 §8，基线 §13）。
 *
 * digest 按 §19.2 计算：clone 后移除 `digest` 自身与 transport signature 字段，
 * RFC 8785 JCS 规范化 + SHA-256，hex 小写并加 `sha256:` 前缀。
 * protocol_version 只认 1.0，未知版本 fail-closed（§5 / §33-6）。
 * actor 仅 buyer|merchant；system/内部事件不得冒充商业角色（§8.1）。
 */

import { contentDigest } from "../jcs.js";
import {
  KNP_PROTOCOL_VERSION,
  NegotiationValidationError,
  requireDigest,
  requireEnum,
  requireIsoTimestamp,
  requireNonEmptyString,
  requireObject,
  requireString,
  schemaError,
} from "./common.js";
import { validateIdentifier } from "./identifiers.js";
import { KNP_ACTIONS, validatePayloadForAction } from "./objects.js";
import type { NegotiationAction, NegotiationActor, NegotiationPayload } from "./objects.js";

/** KNP 未定义的 transport signature 字段名：digest 计算时排除（§19.2）。 */
const TRANSPORT_SIGNATURE_FIELDS = new Set([
  "signature",
  "transport_signature",
  "http_message_signature",
  "x_message_signature",
]);

export type EnvelopeContent = Omit<NegotiationEnvelope, "digest">;

export interface NegotiationEnvelope {
  capability: string;
  protocol_version: typeof KNP_PROTOCOL_VERSION;
  negotiation_id: string;
  exchange_id: string;
  message_id: string;
  in_reply_to?: string;
  actor: NegotiationActor;
  action: NegotiationAction;
  created_at: string;
  payload: NegotiationPayload;
  public_message?: string;
  digest: string;
}

/**
 * 计算 envelope digest（§19.2）：移除 digest 自身与 transport signature 字段，
 * 其余字段 JCS 规范化 + SHA-256。JCS 排序键且跳过 undefined，因此字段顺序
 * 无关、可选字段缺省不影响摘要。
 */
export function computeEnvelopeDigest(input: object): string {
  const clean: Record<string, unknown> = {};
  for (const key of Object.keys(input)) {
    if (key === "digest") continue;
    if (TRANSPORT_SIGNATURE_FIELDS.has(key)) continue;
    clean[key] = (input as Record<string, unknown>)[key];
  }
  return contentDigest(clean);
}

/**
 * 对未签名字段计算 digest 并返回完整 envelope。
 *
 * 命名说明（基线 §19 / §31）：本函数只做内容寻址 digest（JCS + SHA-256），
 * 不做任何密码学签名。KNP/1.0 envelope 的 digest 是完整性/防篡改校验；
 * signature 属于 transport 层且 §19.2 显式从 digest 输入中排除——digest 与
 * signature 是两种性质，故不命名为 sign*。
 */
export function finalizeEnvelope(fields: EnvelopeContent): NegotiationEnvelope {
  return { ...fields, digest: computeEnvelopeDigest(fields) };
}

/** 重算 digest 并与 wire digest 比较；false 表示被篡改或过期。 */
export function verifyEnvelopeDigest(envelope: NegotiationEnvelope): boolean {
  const { digest, ...rest } = envelope;
  return computeEnvelopeDigest(rest) === digest;
}

/**
 * Envelope schema 校验。未知 protocol_version / actor 非 buyer|merchant /
 * action-payload 类型不匹配 / digest 格式错误均 fail-closed。
 * digest 内容一致性由 verifyEnvelopeDigest 单独校验。
 */
export function validateEnvelope(value: unknown): NegotiationEnvelope {
  const obj = requireObject(value, "/");
  const capability = requireNonEmptyString(obj.capability, "/capability");
  const protocolVersion = requireNonEmptyString(obj.protocol_version, "/protocol_version");
  if (protocolVersion !== KNP_PROTOCOL_VERSION) {
    throw new NegotiationValidationError(
      "protocol_version_unsupported",
      `unsupported protocol_version ${protocolVersion}; this runtime implements ${KNP_PROTOCOL_VERSION}`,
      "/protocol_version",
    );
  }
  const negotiationId = validateIdentifier(obj.negotiation_id, "/negotiation_id");
  const exchangeId = validateIdentifier(obj.exchange_id, "/exchange_id");
  const messageId = validateIdentifier(obj.message_id, "/message_id");
  const inReplyTo =
    obj.in_reply_to === undefined ? undefined : validateIdentifier(obj.in_reply_to, "/in_reply_to");
  const actor = requireEnum(obj.actor, ["buyer", "merchant"] as const, "/actor");
  const action = requireEnum(obj.action, KNP_ACTIONS, "/action");
  const createdAt = requireIsoTimestamp(obj.created_at, "/created_at");
  const payload = validatePayloadForAction(action, obj.payload, "/payload");
  const publicMessage =
    obj.public_message === undefined
      ? undefined
      : requireString(obj.public_message, "/public_message");
  const digest = requireDigest(obj.digest, "/digest");

  // §14：clarification_response 必须通过 in_reply_to 引用被回答的澄清消息。
  if (action === "clarification_response" && inReplyTo === undefined) {
    throw schemaError("/in_reply_to", "clarification_response requires in_reply_to");
  }

  return {
    capability,
    protocol_version: KNP_PROTOCOL_VERSION,
    negotiation_id: negotiationId,
    exchange_id: exchangeId,
    message_id: messageId,
    in_reply_to: inReplyTo,
    actor,
    action,
    created_at: createdAt,
    payload,
    public_message: publicMessage,
    digest,
  };
}

File: spec/examples/python/kiwi_ref/envelope.py
"""KNP/1.0 envelope 构造 / digest / 校验（零 kiwi 依赖）。

镜像 Kiwi TS `src/negotiation/domain/envelope.ts` 的语义：

- envelope 必填字段：capability / protocol_version / negotiation_id /
  exchange_id / message_id / actor / action / created_at / payload / digest；
- digest = ``sha256:<hex(JCS(envelope 去掉 digest + transport signature 字段))>``
  （KNP §19.2）；
- ``terms_digest`` = ``content_digest(求值后的 agreed terms)``（§19.3），买卖双方
  各自计算必须一致。

结构校验 fail-closed：无法证明合法即为非法（未知 action / actor / 必填缺失 /
digest 不匹配一律拒绝）。
"""

from __future__ import annotations

import re
import uuid
from typing import Any

from .jcs import content_digest

CAPABILITY = "com.harrylabsj.kiwi.shopping.negotiation"
PROTOCOL_VERSION = "1.0"

ACTORS = frozenset({"buyer", "merchant"})
ACTIONS = frozenset(
    {
        "inquiry",
        "rfq",
        "offer",
        "counter_offer",
        "conditional_offer",
        "clarification",
        "clarification_response",
        "accept_nonbinding",
        "withdraw",
        "decline",
        "cancel",
    }
)

# digest 计算时排除的 transport signature 字段（KNP §19.2，与 TS 一致）。
_SIGNATURE_FIELDS = frozenset(
    {"signature", "transport_signature", "http_message_signature", "x_message_signature"}
)

_DIGEST_RE = re.compile(r"^sha256:[0-9a-f]{64}$")


class EnvelopeError(ValueError):
    """envelope 结构 / digest 校验失败（fail-closed）。"""


def _id(prefix: str) -> str:
    return f"{prefix}_{uuid.uuid4().hex}"


def new_negotiation_id() -> str:
    return _id("neg")


def new_exchange_id() -> str:
    return _id("ex")


def new_message_id() -> str:
    return _id("msg")


def new_offer_id() -> str:
    return _id("off")


def new_agreement_id() -> str:
    return _id("agr")


def _clean_for_digest(envelope: dict) -> dict:
    """克隆并去掉 digest 自身与 transport signature 字段（§19.2）。"""
    return {k: v for k, v in envelope.items() if k not in _SIGNATURE_FIELDS and k != "digest"}


def compute_envelope_digest(fields: dict) -> str:
    return content_digest(_clean_for_digest(fields))


def finalize_envelope(fields: dict) -> dict:
    """对未带 digest 的字段计算 digest，返回完整 envelope。"""
    if "digest" in fields:
        raise EnvelopeError("finalize_envelope: input must not already carry digest")
    return {**fields, "digest": compute_envelope_digest(fields)}


def verify_envelope_digest(envelope: dict) -> bool:
    return compute_envelope_digest(envelope) == envelope.get("digest")


def _require_str(value: Any, path: str) -> str:
    if not isinstance(value, str) or value == "":
        raise EnvelopeError(f"{path} must be a non-empty string")
    return value


def validate_envelope(value: Any) -> dict:
    """结构校验（fail-closed）。返回原 envelope（已校验 digest）。"""
    if not isinstance(value, dict):
        raise EnvelopeError("envelope must be an object")
    if value.get("capability") != CAPABILITY:
        raise EnvelopeError("capability mismatch")
    if value.get("protocol_version") != PROTOCOL_VERSION:
        raise EnvelopeError("protocol_version unsupported")
    for field in ("negotiation_id", "exchange_id", "message_id", "created_at"):
        _require_str(value.get(field), f"/{field}")
    actor = value.get("actor")
    if actor not in ACTORS:
        raise EnvelopeError(f"actor must be one of {sorted(ACTORS)}")
    action = value.get("action")
    if action not in ACTIONS:
        raise EnvelopeError(f"action must be one of {sorted(ACTIONS)}")
    if not isinstance(value.get("payload"), dict):
        raise EnvelopeError("/payload must be an object")
    digest = value.get("digest")
    if not isinstance(digest, str) or _DIGEST_RE.match(digest) is None:
        raise EnvelopeError("/digest must match ^sha256:[0-9a-f]{64}$")
    if not verify_envelope_digest(value):
        raise EnvelopeError("envelope digest does not match content")
    return value


# ---------------------------------------------------------------------------
# 买家动作构造器
# ---------------------------------------------------------------------------


def build_rfq(
    *,
    negotiation_id: str,
    sku: str,
    quantity: int,
    delivery_before: str,
    exchange_id: str | None = None,
    message_id: str | None = None,
    in_reply_to: str | None = None,
    created_at: str,
    actor: str = "buyer",
) -> dict:
    """RFQ envelope（§10）：items + 可选 requested_terms.delivery_before。"""
    fields = {
        "capability": CAPABILITY,
        "protocol_version": PROTOCOL_VERSION,
        "negotiation_id": negotiation_id,
        "exchange_id": exchange_id or new_exchange_id(),
        "message_id": message_id or new_message_id(),
        "actor": actor,
        "action": "rfq",
        "created_at": created_at,
        "payload": {
            "type": "rfq",
            "items": [{"sku": sku, "quantity": {"value": quantity, "unit": "piece"}}],
            "requested_terms": {"delivery_before": delivery_before},
        },
    }
    if in_reply_to is not None:
        fields["in_reply_to"] = in_reply_to
    return finalize_envelope(fields)


def build_counter_offer(
    *,
    negotiation_id: str,
    responding_to_offer_id: str,
    sku: str,
    quantity: int,
    amount_minor: int,
    currency: str = "CNY",
    exchange_id: str | None = None,
    message_id: str | None = None,
    in_reply_to: str,
    created_at: str,
    public_message: str | None = None,
) -> dict:
    """CounterOffer envelope（§12）：完整 proposed_terms（非 patch）。"""
    fields = {
        "capability": CAPABILITY,
        "protocol_version": PROTOCOL_VERSION,
        "negotiation_id": negotiation_id,
        "exchange_id": exchange_id or new_exchange_id(),
        "message_id": message_id or new_message_id(),
        "in_reply_to": in_reply_to,
        "actor": "buyer",
        "action": "counter_offer",
        "created_at": created_at,
        "payload": {
            "type": "counter_offer",
            "offer_id": new_offer_id(),
            "responding_to_offer_id": responding_to_offer_id,
            "proposed_terms": {
                "items": [
                    {
                        "sku": sku,
                        "quantity": {"value": quantity, "unit": "piece"},
                        "unit_price": {"currency": currency, "amount_minor": amount_minor},
                    }
                ]
            },
        },
    }
    if public_message is not None:
        fields["public_message"] = public_message
    return finalize_envelope(fields)


def build_accept_nonbinding(
    *,
    negotiation_id: str,
    accepted_offer_id: str,
    agreed_terms: dict,
    exchange_id: str | None = None,
    message_id: str | None = None,
    in_reply_to: str,
    created_at: str,
) -> dict:
    """AcceptNonbinding envelope（§15）：terms_digest = contentDigest(求值后 terms)。"""
    fields = {
        "capability": CAPABILITY,
        "protocol_version": PROTOCOL_VERSION,
        "negotiation_id": negotiation_id,
        "exchange_id": exchange_id or new_exchange_id(),
        "message_id": message_id or new_message_id(),
        "in_reply_to": in_reply_to,
        "actor": "buyer",
        "action": "accept_nonbinding",
        "created_at": created_at,
        "payload": {
            "type": "accept_nonbinding",
            "offer_id": accepted_offer_id,
            "terms_digest": content_digest(agreed_terms),
        },
    }
    return finalize_envelope(fields)


# ---------------------------------------------------------------------------
# 商家动作构造器
# ---------------------------------------------------------------------------


def offer_terms(
    *,
    sku: str,
    quantity: int,
    amount_minor: int,
    currency: str = "CNY",
    delivery_before: str,
    valid_until: str,
    handoff_destination: str | None = None,
) -> dict:
    """Offer terms（§11）：items 必带 unit_price + fulfillment_terms + valid_until。"""
    terms = {
        "items": [
            {
                "sku": sku,
                "quantity": {"value": quantity, "unit": "piece"},
                "unit_price": {"currency": currency, "amount_minor": amount_minor},
            }
        ],
        "fulfillment_terms": {"delivery_before": delivery_before},
        "valid_until": valid_until,
    }
    if handoff_destination is not None:
        terms["handoff_destination"] = handoff_destination
    return terms


def build_offer(
    *,
    negotiation_id: str,
    in_reply_to: str,
    terms: dict,
    exchange_id: str | None = None,
    message_id: str | None = None,
    created_at: str,
) -> dict:
    """Offer envelope（§11）：RFQ 的回复。"""
    fields = {
        "capability": CAPABILITY,
        "protocol_version": PROTOCOL_VERSION,
        "negotiation_id": negotiation_id,
        "exchange_id": exchange_id or new_exchange_id(),
        "message_id": message_id or new_message_id(),
        "in_reply_to": in_reply_to,
        "actor": "merchant",
        "action": "offer",
        "created_at": created_at,
        "payload": {"type": "offer", "offer_id": new_offer_id(), "terms": terms},
    }
    return finalize_envelope(fields)


def build_conditional_offer(
    *,
    negotiation_id: str,
    in_reply_to: str,
    responding_to_offer_id: str,
    base_terms: dict,
    then_terms: dict,
    quantity_threshold: int = 100,
    exchange_id: str | None = None,
    message_id: str | None = None,
    created_at: str,
) -> dict:
    """ConditionalOffer envelope（§13）：批量条件（aggregate.total_quantity gte 阈值）。"""
    fields = {
        "capability": CAPABILITY,
        "protocol_version": PROTOCOL_VERSION,
        "negotiation_id": negotiation_id,
        "exchange_id": exchange_id or new_exchange_id(),
        "message_id": message_id or new_message_id(),
        "in_reply_to": in_reply_to,
        "actor": "merchant",
        "action": "conditional_offer",
        "created_at": created_at,
        "payload": {
            "type": "conditional_offer",
            "offer_id": new_offer_id(),
            "responding_to_offer_id": responding_to_offer_id,
            "base_terms": base_terms,
            "conditions": [
                {
                    "when": {"all": [{"field": "aggregate.total_quantity", "op": "gte", "value": quantity_threshold}]},
                    "then_terms": then_terms,
                }
            ],
        },
    }
    return finalize_envelope(fields)


def build_agreement(
    *,
    negotiation_id: str,
    accepted_offer_id: str,
    agreed_terms: dict,
    created_at: str,
) -> dict:
    """AcceptedNonbindingAgreement artifact（三副作用恒 false）。"""
    return {
        "type": "accepted_nonbinding_agreement",
        "agreement_id": new_agreement_id(),
        "negotiation_id": negotiation_id,
        "accepted_offer_id": accepted_offer_id,
        "agreed_terms": agreed_terms,
        "terms_digest": content_digest(agreed_terms),
        "accepted_by": ["buyer", "merchant"],
        "created_at": created_at,
        "binding_effect": "nonbinding",
        "creates_order": False,
        "reserves_inventory": False,
        "authorizes_payment": False,
    }
